implementations/rust/ockam/ockam_transport_ble/src/driver/bluetooth_hci/ble_uart.rs

```diff
@@ -0,0 +1,529 @@
+use core::time::Duration;
+
+use bluetooth_hci::event::command::CommandComplete;
+use bluetooth_hci::event::command::ReturnParameters as HciParams;
+use bluetooth_hci::event::command::ReturnParameters::Vendor;
+use bluetooth_hci::host::uart::Hci as _;
+use bluetooth_hci::host::Hci as _;
+
+use bluenrg::event::AttributeHandle;
+use bluenrg::event::BlueNRGEvent;
+use bluenrg::gap::{AuthenticationRequirements, DiscoverableParameters, OutOfBandAuthentication};
+use bluenrg::gatt::{
+    AddCharacteristicParameters, AddServiceParameters, CharacteristicEvent,
+    CharacteristicPermission, CharacteristicProperty, EncryptionKeySize, ServiceType,
+    UpdateCharacteristicValueParameters, Uuid, Uuid::Uuid128,
+};
+use bluenrg::hal::ConfigData;
+use bluenrg::BlueNRG;
+use bluenrg::OwnAddressType;
+
+use bluenrg::event::command::ReturnParameters as Parameters;
+use bluenrg::gap::Commands as GapCommands;
+use bluenrg::gatt::Commands as GattCommands;
+use bluenrg::hal::Commands as HalCommands;
+
+use super::embedded_hal;
+use crate::error::BleError;
+use crate::BleAddr;
+
+type Packet = bluetooth_hci::host::uart::Packet<BlueNRGEvent>;
+pub type Event = bluetooth_hci::event::Event<BlueNRGEvent>;
+
+/// BleContext
+#[derive(Debug, Default)]
+pub struct BleContext {
+    pub service_handle: Option<bluenrg::gatt::ServiceHandle>,
+    pub dev_name_handle: Option<bluenrg::gatt::CharacteristicHandle>,
+    pub appearence_handle: Option<bluenrg::gatt::CharacteristicHandle>,
+
+    pub uart_service_handle: Option<bluenrg::gatt::ServiceHandle>,
+    pub uart_tx_handle: Option<bluenrg::gatt::CharacteristicHandle>,
+    pub uart_rx_handle: Option<bluenrg::gatt::CharacteristicHandle>,
+    pub uart_rx_attribute_handle: Option<AttributeHandle>,
+
+    pub conn_handle: Option<bluenrg::event::ConnectionHandle>,
+}
+
+/// Currently we only support the BLE Peripheral role on embedded
+pub const ROLE: bluenrg::gap::Role = bluenrg::gap::Role::PERIPHERAL;
+
+/// BLE encryption key size
+pub const BLE_ENCRYPTION_KEY_SIZE: usize = 16;
+
+/// If `true`, the `AddCharacteristicParameters::characteristic_value_len`
+/// parameter only takes 1 byte.
+pub const FW_VERSION_BEFORE_V72: bool = true;
+
+/// BLE Uart: Initial Setup (1/5)
+pub fn setup<'buf, SPI, OutputPin1, OutputPin2, InputPin1, GpioError>(
+    spi: &mut SPI,
+    bluetooth: &mut BlueNRG<'buf, SPI, OutputPin1, OutputPin2, InputPin1, GpioError>,
+) -> Result<(), bluenrg::Error<SPI::Error, GpioError>>
+where
+    SPI: blocking::spi::transfer::Default<u8> + blocking::spi::write::Default<u8>,
+    OutputPin1: OutputPin<Error = GpioError>,
+    OutputPin2: OutputPin<Error = GpioError>,
+    InputPin1: InputPin<Error = GpioError>,
+    SPI::Error: Debug,
+    OutputPin1::Error: Debug,
+    OutputPin2::Error: Debug,
+    InputPin1::Error: Debug,
+{
+    debug!("\tsoftware controller reset");
+    block!(bluetooth.with_spi(spi, |controller| { controller.reset() }))?;
+
+    // wait for reset to complete
+    crate::wait_ms!(200);
```

Comment:
Weeellll…

1. Robust according to the bluenrg data sheet + a healthy safety margin. 
2. A little less robust with the `crate::wait_ms!` implementation until we get some `embedded_hal` support for a global timer!

The reason it doesn't make sense to keep spinning here is because the `block!` is on the command having been successfully received by the radio over SPI and not the actual command completion.

Generally hardware resets have no completion response so one just goes with the datasheet interval. :-P